File: backend/app/services/explanation_service.py
import json

import structlog
from openai import AsyncOpenAI

from app.config import settings

logger = structlog.get_logger()

EXPLANATION_SYSTEM_PROMPT = """You are a tax explanation assistant. Your job is to explain ALREADY COMPUTED tax results in plain, friendly language.

Rules:
1. Only explain values that are provided. Do NOT compute or verify any numbers.
2. Use simple language a non-accountant can understand.
3. Be concise but thorough.
4. Reference specific dollar amounts from the results.
5. If there's a refund, explain why in positive terms.
6. If there's a balance owed, explain it matter-of-factly without alarm.
7. Do NOT give tax advice or suggest deductions they didn't take."""


async def generate_explanation(results: dict) -> str:
    """Generate a plain-language explanation of computed tax results."""
    client = AsyncOpenAI(api_key=settings.openai_api_key)

    results_summary = json.dumps(results, indent=2)

    response = await client.chat.completions.create(
        model=settings.openai_model,
        messages=[
            {"role": "system", "content": EXPLANATION_SYSTEM_PROMPT},
            {
                "role": "user",
                "content": f"Explain these tax computation results to the taxpayer:\n\n{results_summary}",
            },
        ],
        temperature=0.5,
        max_tokens=800,
    )

    explanation = response.choices[0].message.content
    logger.info("explanation_generated", length=len(explanation))
    return explanation

File: frontend/src/services/explanation-service.ts
import { openai, OPENAI_MODEL } from "@/lib/openai";

export async function generateExplanation(
  computeResult: Record<string, unknown>,
  factsData: Record<string, unknown>
): Promise<string> {
  const response = await openai.chat.completions.create({
    model: OPENAI_MODEL,
    messages: [
      {
        role: "system",
        content: `You are a tax explanation assistant. Your job is to explain tax computation results in plain, friendly language.

Rules:
1. Reference the actual numbers from the computation.
2. Explain what each major line item means.
3. Do NOT recompute or verify the numbers — the engine is the source of truth.
4. Keep explanations concise (3-5 paragraphs).
5. Mention the filing status and tax year.
6. If there's a refund, explain why. If there's an amount owed, explain why.
7. Do not give tax advice — just explain the numbers.`,
      },
      {
        role: "user",
        content: `Here are the tax computation results to explain:\n\n${JSON.stringify(computeResult, null, 2)}\n\nTax facts used:\n${JSON.stringify(factsData, null, 2)}`,
      },
    ],
    temperature: 0.5,
    max_tokens: 1000,
  });

  return response.choices[0].message.content ?? "Unable to generate explanation.";
}
